cpp/src/io/parquet/experimental/page_index_filter.cu

```diff
@@ -799,178 +808,194 @@ std::vector<std::vector<bool>> aggregate_reader_metadata::compute_data_page_mask
-  if (not has_page_index) {
-    CUDF_LOG_WARN("Encountered missing Parquet page index for one or more output columns");
-    return {};  // An empty data page mask indicates all pages are required
+  // Compute page row offsets and column chunk page offsets for each column
+  auto const num_columns = input_columns.size();
+  std::vector<size_type> page_row_offsets;
+  std::vector<size_type> col_page_offsets;
+  col_page_offsets.reserve(num_columns + 1);
+  col_page_offsets.push_back(0);
+
+  size_type max_page_size = 0;
+
+  if (num_columns <= 2) {
+    std::for_each(
+      column_schema_indices.begin(), column_schema_indices.end(), [&](auto const schema_idx) {
+        auto [col_page_row_offsets, col_max_page_size] =
+          compute_page_row_offsets(per_file_metadata, row_group_indices, schema_idx);
+        page_row_offsets.insert(
+          page_row_offsets.end(), col_page_row_offsets.begin(), col_page_row_offsets.end());
+        max_page_size = std::max<size_type>(max_page_size, col_max_page_size);
+        col_page_offsets.emplace_back(page_row_offsets.size());
+      });
+  } else {
+    auto constexpr max_tasks         = 2;
+    using task_page_row_offsets_type = std::vector<std::pair<std::vector<size_type>, size_type>>;
+    std::vector<std::future<task_page_row_offsets_type>> page_row_offset_tasks{};
+    page_row_offset_tasks.reserve(max_tasks);
+    auto const cols_per_thread = cudf::util::div_rounding_up_unsafe(num_columns, max_tasks);
+
+    // Submit page row offset compute tasks
+    std::transform(thrust::counting_iterator(0),
+                   thrust::counting_iterator(max_tasks),
+                   std::back_inserter(page_row_offset_tasks),
+                   [&](auto const tid) {
+                     return cudf::detail::host_worker_pool().submit_task([&, tid = tid]() {
+                       auto const start_col = std::min(tid * cols_per_thread, num_columns);
+                       auto const end_col   = std::min(start_col + cols_per_thread, num_columns);
+                       task_page_row_offsets_type task_page_row_offsets{};
+                       task_page_row_offsets.reserve(end_col - start_col);
+                       std::transform(
+                         thrust::counting_iterator(start_col),
+                         thrust::counting_iterator(end_col),
+                         std::back_inserter(task_page_row_offsets),
+                         [&](auto const col_idx) {
+                           return compute_page_row_offsets(
+                             per_file_metadata, row_group_indices, column_schema_indices[col_idx]);
+                         });
+                       return task_page_row_offsets;
+                     });
+                   });
+
+    std::for_each(page_row_offset_tasks.begin(), page_row_offset_tasks.end(), [&](auto& task) {
+      auto const& task_page_row_offsets = task.get();
+      for (auto& [col_page_row_offsets, col_max_page_size] : task_page_row_offsets) {
```

Comment:
can these be const?

Reply:
Using `std::make_move_iterator` in the following line now